### ISSUE TITLE
Tensorflow profiling tool does not work on Mac M1 

### ISSUE THREAD

**Issue by chenmoneygithub** [tensorflow/tensorflow#59438]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

source

### Tensorflow Version

2.11

### Custom Code

Yes

### OS Platform and Distribution

M1 Mac

### Python version

3.9

### Current Behaviour?

I cannot do profiling on M1 macbook, but with the same code I can profile with linux machine (tried on colab).

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
from datetime import datetime

cifar = tf.keras.datasets.cifar100
(x_train, y_train), (x_test, y_test) = cifar.load_data()
model = tf.keras.applications.ResNet50(
    include_top=True,
    weights=None,
    input_shape=(32, 32, 3),
    classes=100,
)

logs = "logs/" + datetime.now().strftime("%Y%m%d-%H%M%S")

tboard_callback = tf.keras.callbacks.TensorBoard(
    log_dir=logs, histogram_freq=1, profile_batch=(50, 150)
)

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
optimizer = tf.keras.optimizers.Adam()
legacy_optimizer = tf.keras.optimizers.legacy.Adam()
# switch optimizer to test performance
model.compile(optimizer=optimizer, loss=loss_fn, metrics=["accuracy"])

model.fit(
    x_train,
    y_train,
    epochs=2,
    batch_size=16,
    steps_per_epoch=200,
    callbacks=[tboard_callback],
)
```
```


### Relevant log output

_No response_</details>

**Comment by chenmoneygithub** [comment]
On M1 the profiling outputs: ![m1_profile](https://user-images.githubusercontent.com/22925031/214459424-33402cf3-8df1-4195-998c-893f998fbf2e.png)

On colab the profiling outputs: ![colab_profile](https://user-images.githubusercontent.com/22925031/214459430-3ac7e9be-de63-453b-81b0-1f6e2e9898ef.png)

**Comment by learning-to-play** [comment]
Hi @kulinseth, Could you please share your thoughts?

**Comment by mohantym** [comment]
Hi @chenmoneygithub !
Thanks for sharing your observation on profiling tool w.r.to Mac M1.

Below is the output in my local M1 machine after going with Metal plugin instructions with Python 3.9 (not working with Python 3.8 now)

```
2023-01-27 17:56:02.193049: I tensorflow/core/profiler/lib/profiler_session.cc:101] Profiler session initializing.
2023-01-27 17:56:02.193072: I tensorflow/core/profiler/lib/profiler_session.cc:116] Profiler session started.
2023-01-27 17:56:02.193117: I tensorflow/core/profiler/lib/profiler_session.cc:128] Profiler session tear down.
2023-01-27 17:56:02.413295: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
Epoch 1/2
/Users/mohantym/miniforge3/envs/mac-test/lib/python3.9/site-packages/keras/backend.py:5585: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(
 49/200 [======>.......................] - ETA: 26s - loss: 7.3209 - accuracy: 0.01792023-01-27 17:56:19.055651: I tensorflow/core/profiler/lib/profiler_session.cc:101] Profiler session initializing.
2023-01-27 17:56:19.055664: I tensorflow/core/profiler/lib/profiler_session.cc:116] Profiler session started.
149/200 [=====================>........] - ETA: 9s - loss: 6.5248 - accuracy: 0.01802023-01-27 17:56:37.421900: I tensorflow/core/profiler/lib/profiler_session.cc:67] Profiler session collecting data.
2023-01-27 17:56:37.778513: I tensorflow/core/profiler/lib/profiler_session.cc:128] Profiler session tear down.
2023-01-27 17:56:37.780399: I tensorflow/core/profiler/rpc/client/save_profile.cc:164] Collecting XSpace to repository: logs/20230127-175602/plugins/profile/2023_01_27_17_56_37/mohantym-macbookpro2.roam.internal.xplane.pb
200/200 [==============================] - 46s 192ms/step - loss: 6.4378 - accuracy: 0.0191
Epoch 2/2
200/200 [==============================] - 41s 207ms/step - loss: 5.9023 - accuracy: 0.0244
```
Thank you!

**Comment by chenmoneygithub** [comment]
@mohantym Thanks! I have the same profiling output, and it does generate a file. However, if you check the profile results on Tensorboard you will see nothing gets really traced, please see the screenshot in  [this comment](https://github.com/tensorflow/tensorflow/issues/59438#issuecomment-1402919577)

**Comment by kulinseth** [comment]
Hi @mohantym and @chenmoneygithub, thanks for reaching out and your interest . Currently we haven’t implemented the pluggable profiler api for Metal plug-in . Is this something you are interested in?

**Comment by mohantym** [comment]
Hi @kulinseth !

Thanks for the replying . I only shared my observations and agree with your point .
It would be great to have profiling option enabled in metal-plugin side of Tensorflow . 

Thank you!

**Comment by chenmoneygithub** [comment]
@kulinseth Thx! We (Keras team) don't have the knowledge on how to make the profiling tool. Will any team in Apple have the bandwidth for it?

**Comment by chenmoneygithub** [comment]
@kulinseth Sorry I forgot to include the context for it. Currently we are seeing an issue that the new Keras optimizer runs very slowly on M1 mac, so we need the profiling tool to do a deep dive. Your help will be very much appreciated! We are struggling now without the visual tool.

**Comment by kulinseth** [comment]
> @kulinseth Sorry I forgot to include the context for it. Currently we are seeing an issue that the new Keras optimizer runs very slowly on M1 mac, so we need the profiling tool to do a deep dive. Your help will be very much appreciated! We are struggling now without the visual tool.

Thanks @chenmoneygithub , that makes sense. We noticed this as well. We are seeing upto 3x and more slowdown in the Keras optimizer API. With the new optimizer API in Keras we are falling back to using Python layer to call into the GPU mapped operations to implement the Optimizer layer instead of the `ResourceApplyAdam()` implementation of the optimizer in the Metal plugin.  

Can you check one thing, if you use the Legacy Optimizer op does it work and if the performance is fast?

```
import tensorflow as tf
from tensorflow.keras.optimizers.legacy import Adam

cifar = tf.keras.datasets.cifar100
(x_train, y_train), (x_test, y_test) = cifar.load_data()
model = tf.keras.applications.ResNet50(
    include_top=True,
    weights=None,
    input_shape=(32, 32, 3),
    classes=100,)

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer=Adam(), loss=loss_fn, metrics=["accuracy"])
model.fit(x_train, y_train, epochs=5, batch_size=64)
```

**Comment by chenmoneygithub** [comment]
@kulinseth Yea, the legacy optimizer is faster in my comparison, and the scenario only happens on M1 mac. Rostam has invited me to the meeting with you tomorrow, we can discuss more details there, thanks!

**Comment by kulinseth** [comment]
> @kulinseth Yea, the legacy optimizer is faster in my comparison, and the scenario only happens on M1 mac. Rostam has invited me to the meeting with you tomorrow, we can discuss more details there, thanks!

Thanks! @chenmoneygithub

**Comment by chenmoneygithub** [comment]
@kulinseth Kulin, we have made a change to automatically convert optimizer to legacy if arm Mac is detected (for model.fit() only). Please let me know if there is new findings about the speed drop on your side, thanks!

**Comment by kulinseth** [comment]
@chenmoneygithub thanks for much. I will update here with the findings. Can you please share the commit id? Also will it be part of the 2.12 release?
